### PR TITLE
[#386] add retries to ModelClient

### LINKED ISSUE #386: Add retries to model client

### PR DESCRIPTION
**Pull request by botsman**
This closes #386 

Add retries to requests in ModelClient

### CODE REVIEW

**Comment by botsman** [comment]
bot, test all

**Comment by botsman** [comment]
bot, test all

**Comment by botsman** [comment]
bot, test all

**Comment by botsman** [comment]
bot, test app

**Comment by botsman** [comment]
bot, test app

**Comment by botsman** [comment]
bot, test all

**Comment by botsman** [comment]
bot, test all

**Comment by alexander-semenets** [comment]
bot, test all

File: legion/legion/model/client.py [hunk 1]
```diff
@@ -82,6 +82,13 @@ def __init__(self, model_id, model_version, token=None, host=None, http_client=N
         else:
             self._http_client = requests
 
+        if http_client and http_exception:
```
Comment (mcdoker18) [inline legion/legion/model/client.py]: Please, explain so difficult logic

File: legion/legion/model/client.py [hunk 2]
```diff
@@ -225,6 +232,39 @@ def _additional_kwargs(self):
...
+        Send request with provided method and other parameters
+        :param http_method: HTTP method
+        :type http_method: str
+        :param url: url to send request to
+        :type url: str
+        :param data: request data
+        :type data: any
+        :param files: files
+        :type files: dict
+        :return: list -- parsed model response
```
Comment (mcdoker18) [inline legion/legion/model/client.py]: dict?

File: legion/legion/model/client.py [hunk 3]
```diff
@@ -225,6 +232,39 @@ def _additional_kwargs(self):
...
+        client_method = getattr(self._http_client, http_method)
+
+        if data:
+            kwargs['data'] = data
+        if files:
+            kwargs['files'] = files
+
+        def check_function():
+            try:
+                return client_method(url, **kwargs)
```
Comment (mcdoker18) [inline legion/legion/model/client.py:260]: Please, write a unit test for the retries

**Review by mcdoker18 — CHANGES_REQUESTED** [review]

**Comment by botsman** [comment]
bot, test all

File: legion/tests/test_model_client.py [hunk 4]
```diff
@@ -64,6 +65,16 @@ def test_client_building_from_env(self):
             self.assertEqual(client.info_url, root_url + '/info')
             self.assertEqual(client.build_invoke_url(), root_url + '/invoke')
 
+    def test_client_request_retry_on_error(self):
+        with patch('requests.get') as mock:
```
Comment (mcdoker18) [inline legion/tests/test_model_client.py]: Maybe we will use the `responses`  library?

File: legion/tests/test_model_client.py [hunk 5]
```diff
@@ -64,6 +65,16 @@ def test_client_building_from_env(self):
...
+            mock.side_effect = requests.RequestException('Some exception')
+            client = legion.model.client.ModelClient(self.MODEL_ID, self.MODEL_VERSION)
+            try:
+                client._request('get', 'http://some_url')
+            except requests.RequestException:
```
Comment (mcdoker18) [inline legion/tests/test_model_client.py]: Please, use https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises

**Review by mcdoker18 — CHANGES_REQUESTED** [review]

**Comment by botsman** [comment]
bot, test all

File: legion/legion/model/client.py [hunk 6]
```diff
@@ -48,8 +48,8 @@ class ModelClient:
     Model HTTP client
     """
 
-    def __init__(self, model_id, model_version, token=None, host=None, http_client=None, use_relative_url=False,
-                 timeout=None):
+    def __init__(self, model_id, model_version, token=None, host=None, http_client=None, http_exception=None,
+                 use_relative_url=False, timeout=None):
         """
         Build client
```
Comment (kirillmakhonin) [inline legion/legion/model/client.py:54]: You missed adding docs for new arguments

**Review by kirillmakhonin — CHANGES_REQUESTED** [review]

File: legion/legion/model/client.py [hunk 7]
```diff
@@ -48,8 +48,8 @@ class ModelClient:
     Model HTTP client
     """
 
-    def __init__(self, model_id, model_version, token=None, host=None, http_client=None, use_relative_url=False,
-                 timeout=None):
+    def __init__(self, model_id, model_version, token=None, host=None, http_client=None, http_exception=None,
```
Comment (kirillmakhonin) [inline legion/legion/model/client.py]: Maybe we can use request's exception by default?

Comment (botsman) [reply legion/legion/model/client.py]: Fixed

File: legion/legion/model/client.py (hunk 6 above)
Comment (botsman) [reply legion/legion/model/client.py:54]: Fixed

**Comment by botsman** [comment]
bot, test all

**Review by alexander-semenets — APPROVED** [review]

**Review by mcdoker18 — APPROVED** [review]

**Review by kirillmakhonin — APPROVED** [review]

**Pull request merged**